### PR TITLE
Feature/26 trains

### LINKED ISSUE #26: Implement interface for a `train`
## User stories
As a Developer I want to have an object representing a train in order to influence the simulation by manipulating the train.

## Dependencies
Depends on: 
 - #5 
 - #15 

## (Wiki) References
- See #10 

## Acceptance Criteria
- [x] `get(/set)` current train position
- [x] `get/set` current train route
- [x] `get` current edge
- [ ] ~~`get` current fahrstrasse~~
- [x] `get` train type
- [x] `get/set` timetable
- [ ] ~~`get` blocksections~~
- [x] `get/set` priority
- [x] `spawn()` to add a train to the simulation

### PR DESCRIPTION
**Pull request by SaturnHafen**
Fixes #26 
Description (what might a PO or reviewer want to know?)

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [ ] Docs (code, wiki & diagrams) updated
- [ ] Breaking changes anounced
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

**Comment by SaturnHafen** [comment]
Waiting for #116

File: src/wrapper/simulation_objects.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        """Updates the timetable of this train to the given timetable
+
+        :param timetable: the new timetable
+        """
+        self._timetable = timetable
+
+    def __init__(
+        self,
+        identifier: str = None,
+        timetable: List[Platform] = None,
```
Comment (CR1337) [inline src/wrapper/simulation_objects.py]: The Spawner needs your `__init__` to accept a list of platform ids.
Sorry 😑

**Review by CR1337 — CHANGES_REQUESTED** [review]
Look like a lot of work. Good job :+1: 

We just need one change: the platform list in `__init__`

File: src/wrapper/simulation_objects.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    @property
+    def route(self) -> str:
+        """This method returns the current sumo-route-id.
+
+        :return: The route this vehicle is following
+        """
+        return self._route
+
+    @route.setter
+    def route(self, route: int) -> None:
```
Comment (SaturnHafen) [inline src/wrapper/simulation_objects.py]:
```suggestion
    def route(self, route_id: str) -> None:
```

File: src/wrapper/simulation_objects.py [hunk 3]
```diff
@@ -84,3 +86,222 @@ def update(self, data: dict):
...
+    _max_speed: float = None
+    _priority: int = None
+    _train_type: str = None
+    _timetable: List[Platform] = None
+
+    @property
+    def track(self) -> Track:
+        """Returns the current track the train is on
+
+        :return: _description_
```
Comment (Lietze) [inline src/wrapper/simulation_objects.py]:
```suggestion
        :return: The current track the train is on
```
I think you missed this

File: src/wrapper/simulation_objects.py [hunk 4]
```diff
@@ -84,3 +86,222 @@ def update(self, data: dict):
...
+
+class Train(SimulationObject):
+    """A train driving around in the simulation."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    _position: Tuple[float, float] = None
+    _route: str = None
+    _track: Track = None
+    _vehicle_type: str = None
```
Comment (Lietze) [inline src/wrapper/simulation_objects.py]: Why is vehicle_type needed? It is not used anywhere and seems like a duplicate of train_type.

File: src/wrapper/simulation_objects.py [hunk 5]
```diff
@@ -84,3 +86,222 @@ def update(self, data: dict):
...
+    track: Track
+    station_name: str
+    platform_number: int
+    blocked: bool
+
+
+class Train(SimulationObject):
+    """A train driving around in the simulation."""
+
+    # pylint: disable=too-many-instance-attributes
```
Comment (Lietze) [inline src/wrapper/simulation_objects.py]: Maybe it is possible to move some grouped instance attributes into a helper class, so that this is not necessary anymore, and it is easier to understand quickly? 
I thought about a class TrainType, which has the fields `_max_speed`, `_priority` and `_train_type`.

File: tests/wrapper/test_simulation_objects.py [hunk 6]
```diff
@@ -28,3 +28,132 @@ def test_update_state(self, traffic_update):
...
+        )
+
+    def test_speed(self, train):
+        assert train.speed == 10.2
+
+    def test_route(self, train):
+        assert train.route == "testing-route"
+
+    def test_set_route(self, train, vehicle_route):
+        # pylint: disable=unused-argument
```
Comment (Lietze) [inline tests/wrapper/test_simulation_objects.py:104]: Why is this needed? Or why does the vehicle_route need to be an argument?

File: tests/wrapper/test_simulation_objects.py [hunk 7]
```diff
@@ -28,3 +28,132 @@ def test_update_state(self, traffic_update):
...
+        train.max_speed = 100
+        assert train.max_speed == 100
+
+    def test_priority(self, train):
+        assert train.priority == 0
+
+    def test_set_priority(self, train):
+        train.priority = 10
+        assert train.priority == 10
```
Comment (Lietze) [inline tests/wrapper/test_simulation_objects.py]: Some of these test look rather trivial, but I guess that is better than not having them.

**Review by Lietze — COMMENTED** [review]
This looks great! I found a few things I need clarification of and one little fix, but this seems like it could actually work with SUMO ;)

File: tests/wrapper/test_simulation_objects.py (hunk 6 above)
Comment (SaturnHafen) [reply tests/wrapper/test_simulation_objects.py:104]: `vehicle_route` references the fixture declared [here](https://github.com/BP2022-AP1/bp2022-ap1/pull/66/files/04f88cc6d09698686aff7714202bbeeec940482a#diff-7c747795f1156f37736b6c63a776de5e8aee451c9d700a29d6ad22e25504747dR60). This fixture is mocking the `vehicle.setRouteID` from traci so that we don't have to start sumo in order to run the tests. We need to have the fixture in scope (as an argument), but don't use the return value so the linter-ignore is justified.

File: src/wrapper/simulation_objects.py (hunk 4 above)
Comment (SaturnHafen) [reply src/wrapper/simulation_objects.py]: good catch; removed!

File: src/wrapper/simulation_objects.py (hunk 5 above)
Comment (SaturnHafen) [reply src/wrapper/simulation_objects.py]: I added a helper class

**Review by Lietze — APPROVED** [review]
Great job! I'm looking forward to using this.

**Review by CR1337 — APPROVED** [review]
As far as the `Train` - `Schedule` interaction is concerned it looks fine to me. Good work :)
